cmd/tvx/main.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/filecoin-project/go-jsonrpc"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+
+	"github.com/urfave/cli/v2"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/client"
+)
+
+var apiEndpoint string
+
+var apiFlag = cli.StringFlag{
+	Name: "api",
+	Usage: "json-rpc api endpoint, formatted as [token]:multiaddr;" +
+		"tvx uses unpriviliged operations, so the token may be omitted," +
+		"but permissions may change in the future",
+	EnvVars:     []string{"FULLNODE_API_INFO"},
+	DefaultText: "",
+	Destination: &apiEndpoint,
+}
+
+func main() {
+	app := &cli.App{
+		Name: "tvx",
+		Description: `tvx is a tool for extracting and executing test vectors. It has two subcommands.
+
+   tvx extract extracts a test vector from a live network. It requires access to
+   a Filecoin client that exposes the standard JSON-RPC API endpoint. Set the API
+   endpoint on the FULLNODE_API_INFO env variable, or through the --api flag. The
+   format is token:multiaddr. Only message class test vectors are supported
+   for now.
+
+   tvx exec executes test vectors against Lotus. Either you can supply one in a
+   file, or many as an ndjson stdin stream.`,
+		Usage: "tvx is a tool for extracting and executing test vectors",
+		Commands: []*cli.Command{
+			extractCmd,
+			execCmd,
+		},
+	}
+
+	sort.Sort(cli.CommandsByName(app.Commands))
+	for _, c := range app.Commands {
+		sort.Sort(cli.FlagsByName(c.Flags))
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func makeAPIClient() (api.FullNode, jsonrpc.ClientCloser, error) {
```

Comment:
This should just use the methods from `cli/`:
* Add a global repo flag to cli.App above:
```
			&cli.StringFlag{
				Name:    "repo",
				EnvVars: []string{"LOTUS_PATH"},
				Hidden:  true,
				Value:   "~/.lotus", // TODO: Consider XDG_DATA_HOME
			},
```
* In any commands which needs the API:
```
lcli "github.com/filecoin-project/lotus/cli"
...
		api, closer, err := lcli.GetFullNodeAPI(cctx)
		if err != nil {
			return err
		}

		defer closer()
		ctx := lcli.ReqContext(cctx)
```

Reply:
I can add this, which can be useful if you're running the binary from the same machine as the client, that is not my case, for example ;-)

Reply:
Caught up with @magik6k 1:1. I've dropped the `--api` flag (now only an env var), and added the `--repo` flag and env variable binding. We now use the lotus CLI lib, and the usage docs on the command document how the API endpoint is resolved.